dali/kernels/reduce/reduce_drop_dims.h

```diff
@@ -0,0 +1,245 @@
+ * cannot be used to index the the tensor of means, which has more reduced dimensions.
+ * Example - calculating mean/stddev of all pixel brightness values in a mutli-channel video:
+ * input layout FHWC
+ * output layout 1HW1
+ * When calculating variance in F, we still see HWC as the (fused) inner dimension, but it is
+ * different than HW1 in the tensor of means. This class implements this kind of reindexing.
+ *
+ * The reindexing is done by either dividing and multiplying by old/new strides or by takind modulo.
+ */
+struct DropDims {
+  static constexpr int kMaxDims = 4;
+
+  int64_t div[kMaxDims];
+  int64_t mul[kMaxDims];
+  int64_t mod[kMaxDims];
+  int start = 2 * kMaxDims;
+
+  DropDims() = default;
+
+  /**
+   * Collapses adjacent groups of reduced/non-reduced dimensions.
+   * The input can have up to 64 dimensions, however, the result must havfe
+   * at most 8 groups of reduced/non-reduced dimensions.
+   *
+   * Example:
+   * ```
+   * in_shape:  [ 2, 3, 4, 5, 6 ]
+   * reduced:     ^        ^  ^
+   * axis_mask = 0b11001  (looks reversed - LSB is dim 0)
+   *
+   * out_shape: [ 2, 12, 30 ]
+   * reduced:     ^       ^
+   * out_mask = 0b101
+   * ```
+   */
+  template <typename Indices>
+  static int simplify(int64_t *out_shape, unsigned &out_mask,
+                      const Indices &in_shape, uint64_t axis_mask) {
+    int dims = size(in_shape);
+    int d = 0;
+    out_shape[0] = in_shape[0];
+    bool prev = axis_mask & 1;
+    out_mask = prev ? 1u : 0u;
+    for (int i = 1; i < dims; i++) {
+      if (in_shape[i] == 1)
+        continue;
+      bool flag = (axis_mask >> i) & 1;
+      if (flag != prev) {
+        d++;
+        if (d > 2*kMaxDims)
+          throw std::out_of_range("Maximum number of dimension groups exceeded");
+        out_shape[d] = in_shape[i];
+        out_mask |= (flag ? 1u : 0u) << d;
+      } else {
+        out_shape[d] *= in_shape[i];
+      }
+      prev = flag;
+    }
+    d++;
+    if (d > 2*kMaxDims)
+      throw std::out_of_range("Maximum number of dimension groups exceeded");
+
+    return d;
+  }
+
+  /**
+   * @brief Initializes reindexing given shape and mask.
+   */
+  template <typename Indices>
+  DropDims(const Indices &in_shape, uint64_t reduced_axes) {
+    int64_t shape[kMaxDims];
+    unsigned axis_mask;
+    int d = simplify(shape, axis_mask, in_shape, reduced_axes);
+    start = 2 * kMaxDims;
+
+    if (d == 1) {
+      // short circuit trivial case
+      if (axis_mask == 1)
+        start = -1;
+      return;
+    }
+
+    int nmod = 0;
+    int ndiv = 0;
+
+    int64_t volumes[2*kMaxDims];
+    int64_t kept_volumes[2*kMaxDims];
+    int64_t vol_total = 1, vol_kept = 1;
+    int reduced_dims = 0;
+    int kept_dims = 0;
+    for (int i = d - 1; i >= 0; i--) {
+      volumes[i] = vol_total;
+      kept_volumes[i] = vol_kept;
+      vol_total *= shape[i];
+      if ((axis_mask & (1 << i)) == 0) {
```

Comment:
Why it differentiates between `b101` and `b010`?

Reply:
HWC and mask 101 is reduced to 1W1; HWC and mask 010 is reduced to H1C